src/main/java/net/oauth/jsontoken/JsonToken.java

```diff
@@ -329,7 +342,9 @@ private String getSignature() throws SignatureException {
     }
 
     if (signer == null) {
-      throw new SignatureException("can't sign JsonToken with signer.");
+      throw new SignatureException(
+          "can't sign JsonToken with signer",
+          new InvalidJsonTokenException(ErrorCode.INTERNAL_ERROR));
```

Comment:
I think IllegalStateException makes more sense here than InvalidJsonTokenException. It's an incorrect use of this library to try to sign a jwt without providing a signer.

Reply:
Changed.

Reply:
To ensure that IllegalStateException is thrown in the async parser in this scenario, we can introduce an additional custom package-private exception specifically for rethrowing another exception. Unlike the InvalidJsonTokenException, this exception wrapper would be more general (so it would work in any scenarios) and is only intended to be used within the library for message passing. Whenever we encounter this wrapper execution, we can simply throw whatever is underneath. 

With this approach, the flow might go something like: 
1. Check if the cause is an InvalidJsonTokenException. If so, rethrow it.
2. Check if the cause is this new custom wrapper. If so, rethrow whatever exception is wrapped.
3. Check if the exception itself is a checked exception. If so, throw an InvalidJsonTokenException with an unknown error code.
4. Finally, just rethrow the exception (the remaining exceptions would just be runtime exceptions).

An upside to this is that we can have clear distinction between SignatureException (or any other exception) from something like this and a failed checker. The downside would be it adds more layers of wrappings and seems a bit hacky.

eg. 
throw new SignatureException(
  "some error message",
  new RethrowableException(
    new IllegalStateException()
  ))

Reply:
What about an ILLEGAL_STATE error code which gets translated to IllegalStateException in AsyncJsonTokenParser? So we still have the clear distinction without the extra layer of wrapping?

Reply:
I have introduced the ILLEGAL_STATE error code with a description of what would happen if it is seen.